pkg/networkdriver/driver.go

```diff
@@ -0,0 +1,303 @@
+			}
+
+			driver.logger.DebugContext(ctx, "enabled manager", logfields.Type, manager)
+			driver.deviceManagers[manager] = d
+		}
+
+		driver.logger.DebugContext(ctx,
+			"starting driver with config",
+			logfields.Config, driver.config)
+
+		// create path for our driver plugin socket.
+		err := os.MkdirAll(driverPluginPath(driver.driverName), 0750)
+		if err != nil {
+			return fmt.Errorf("failed to create plugin path %s: %w", driverPluginPath(driver.driverName), err)
+		}
+
+		pluginOpts := []kubeletplugin.Option{
+			kubeletplugin.DriverName(driver.driverName),
+			kubeletplugin.NodeName(node_types.GetName()),
+			kubeletplugin.KubeClient(driver.kubeClient),
+		}
+
+		p, err := kubeletplugin.Start(ctx, driver, pluginOpts...)
+		if err != nil {
+			return err
+		}
+
+		driver.draPlugin = p
+
+		err = wait.PollUntilContextTimeout(
+			ctx, driver.config.DraRegistrationRetry, driver.config.DraRegistrationTimeout, true,
+			func(context.Context) (bool, error) {
+				registrationStatus := driver.draPlugin.RegistrationStatus()
+				if registrationStatus == nil {
+					return false, nil
+				}
+
+				driver.logger.DebugContext(
+					ctx, "DRA registration status",
+					logfields.Status, registrationStatus,
+				)
+
+				return registrationStatus.PluginRegistered, nil
+			})
+
+		if err != nil {
+			return fmt.Errorf("DRA plugin registration failed: %w", err)
+		}
+
+		driver.logger.DebugContext(ctx,
+			"DRA plugin registration successful",
+			logfields.DriverName, driver.driverName,
+		)
+
+		// register the NRI plugin
+		nriOptions := []stub.Option{
+			stub.WithPluginName(driver.driverName),
+			stub.WithPluginIdx("00"),
+			// https://github.com/containerd/nri/pull/173
+			// Otherwise it silently exits the program
+			stub.WithOnClose(func() {
+				driver.logger.WarnContext(
+					ctx, "NRI plugin closed",
+					logfields.DriverName, driver.driverName,
+				)
+			}),
+		}
+
+		nriStub, err := stub.New(driver, nriOptions...)
+		if err != nil {
+			return fmt.Errorf("failed to create plugin stub: %w", err)
+		}
+
+		driver.nriPlugin = nriStub
+
+		driver.jg.Add(job.OneShot("networkdriver-nri-plugin-run", func(ctx context.Context, _ cell.Health) error {
+			for {
+				if err := driver.nriPlugin.Run(ctx); err != nil {
+					driver.logger.ErrorContext(
+						ctx, "NRI plugin failed",
+						logfields.Error, err,
+						logfields.Name, driver.driverName,
+					)
+				}
+				select {
+				case <-ctx.Done():
+					return nil
+				default:
+					driver.logger.DebugContext(ctx, "Restarting NRI plugin", logfields.Name, driver.driverName)
+				}
+			}
+		}))
+
+		driver.jg.Add(job.Timer("networkdriver-dra-publish-resources", driver.publish, driver.config.PublishInterval))
+		return driver.publish(ctx)
+	}))
+
+	return nil
+}
+
+// Stop stops the nri and dra hooks.
+func (driver *Driver) Stop(ctx cell.HookContext) error {
+	driver.logger.DebugContext(ctx, "Stopping network driver...")
+
+	// Stop NRI plugin first
+	if driver.nriPlugin != nil {
```

Comment:
Nit: no need to check for `nil`, if we fail to create the NRI stub in the `Start` hook we'll return with an error and that will stop the hive construction process (IOW we'll never arrive here).
Same for`driver.draPlugin`.

Reply:
right, but that is a matter of ordering - and in case we eventually change the startup order, we'd not be properly stopping the registrations.
i'd keep it all as it is to make sure that the Stop() call never misses any of them